### ISSUE TITLE
Issue with the different representations of an arity 2 H box leading to radically different matrices

### ISSUE THREAD

**Issue by RDPEast** [Quantomatic/pyzx#56]
The following seems to occur when studying the following diagrams
![image](https://user-images.githubusercontent.com/13621329/96365585-fcfe4200-1141-11eb-8ac7-5a86ad5cc404.png)

The first is correct, the second is nonsense.

The Json for the first is here

spinjson = '{"wire_vertices": {"b0": {"annotation": {"boundary": true, "coord": [0.45399999999999974, -1.010875], "input": true, "output": false}}, "b1": {"annotation": {"boundary": true, "coord": [0.8039999999999998, -9.410874999999999], "input": true, "output": false}}, "b2": {"annotation": {"boundary": true, "coord": [21.9, -0.6000000000000001], "input": false, "output": true}}, "b3": {"annotation": {"boundary": true, "coord": [22.1, -10.9], "input": false, "output": true}}}, "node_vertices": {"v0": {"annotation": {"coord": [5.415683593750001, -1.0614531250000003]}, "data": {"type": "Z"}}, "v1": {"annotation": {"coord": [15.778683593750003, -1.0313281250000004]}, "data": {"type": "Z"}}, "v2": {"annotation": {"coord": [4.542058593750001, -9.526578125000002]}, "data": {"type": "Z"}}, "v3": {"annotation": {"coord": [16.200433593750002, -9.978453125000001]}, "data": {"type": "Z"}}, "v4": {"annotation": {"coord": [6.741183593750001, -5.068078125000001]}, "data": {"type": "hadamard", "is_edge": "false", "value": "\\\\pi"}}, "v5": {"annotation": {"coord": [13.218058593750001, -5.0379531250000005]}, "data": {"type": "hadamard", "is_edge": "false", "value": "\\\\pi"}}, "v8": {"annotation": {"coord": [9.96455859375, -2.085703125]}, "data": {"type": "hadamard", "is_edge": "false", "value": "\\\\pi"}}, "v9": {"annotation": {"coord": [10.265808593750002, -7.598578125000001]}, "data": {"type": "hadamard", "is_edge": "false", "value": "\\\\pi"}}, "v10": {"annotation": {"coord": [10.386308593750002, -5.339203125000001]}, "data": {"type": "Z"}}, "v20": {"annotation": {"coord": [2.7948085937500005, -8.532453125000002]}, "data": {"type": "Z"}}, "v6": {"annotation": {"coord": [4.270933593750001, -3.531703125000001]}, "data": {"type": "X", "value": "\\\\pi"}}, "v7": {"annotation": {"coord": [1.5898085937500004, -3.8932031250000003]}, "data": {"type": "hadamard", "is_edge": "false", "value": "\\\\pi/6"}}, "v11": {"annotation": {"coord": [1.7404335937500002, -5.670578125000001]}, "data": {"type": "hadamard", "is_edge": "false", "value": "11\\\\pi/6"}}, "v12": {"annotation": {"coord": [2.7948085937500005, -4.586078125000001]}, "data": {"type": "Z", "value": "\\\\pi/2"}}, "v13": {"annotation": {"coord": [0.5354335937500001, -4.736703125000001]}, "data": {"type": "X", "value": "\\\\pi"}}, "v14": {"annotation": {"coord": [3.0960585937500005, -6.604453125000001]}, "data": {"type": "hadamard", "is_edge": "false", "value": "\\\\pi"}}, "v15": {"annotation": {"coord": [3.8491835937500003, -8.411953125]}, "data": {"type": "Z"}}, "v16": {"annotation": {"coord": [3.969683593750001, -6.303203125000001]}, "data": {"type": "X", "value": "\\\\pi"}}, "v17": {"annotation": {"coord": [10.597183593750001, -1.0463906250000004]}, "data": {"type": "hadamard", "is_edge": "true"}}, "v18": {"annotation": {"coord": [18.839341796875, -0.8156640625000002]}, "data": {"type": "hadamard", "is_edge": "true"}}, "v19": {"annotation": {"coord": [10.371246093750003, -9.752515625000001]}, "data": {"type": "hadamard", "is_edge": "true"}}, "v21": {"annotation": {"coord": [2.6730292968750007, -9.4687265625]}, "data": {"type": "hadamard", "is_edge": "true"}}}, "undir_edges": {"e0": {"src": "v0", "tgt": "b0"}, "e1": {"src": "v0", "tgt": "v4"}, "e2": {"src": "v0", "tgt": "v8"}, "e3": {"src": "v0", "tgt": "v17"}, "e4": {"src": "v1", "tgt": "v17"}, "e5": {"src": "v1", "tgt": "v5"}, "e6": {"src": "v1", "tgt": "v8"}, "e7": {"src": "v1", "tgt": "v18"}, "e8": {"src": "b2", "tgt": "v18"}, "e9": {"src": "v2", "tgt": "v4"}, "e10": {"src": "v2", "tgt": "v9"}, "e11": {"src": "v2", "tgt": "v19"}, "e12": {"src": "v3", "tgt": "v19"}, "e13": {"src": "v2", "tgt": "v21"}, "e14": {"src": "b1", "tgt": "v21"}, "e15": {"src": "v3", "tgt": "b3"}, "e16": {"src": "v3", "tgt": "v5"}, "e17": {"src": "v3", "tgt": "v9"}, "e18": {"src": "v4", "tgt": "v10"}, "e19": {"src": "v5", "tgt": "v10"}, "e20": {"src": "v8", "tgt": "v10"}, "e21": {"src": "v9", "tgt": "v10"}, "e22": {"src": "v10", "tgt": "v6"}, "e23": {"src": "v6", "tgt": "v12"}, "e24": {"src": "v7", "tgt": "v13"}, "e25": {"src": "v7", "tgt": "v12"}, "e26": {"src": "v11", "tgt": "v13"}, "e27": {"src": "v11", "tgt": "v12"}, "e28": {"src": "v12", "tgt": "v16"}, "e29": {"src": "v14", "tgt": "v20"}, "e30": {"src": "v14", "tgt": "v15"}, "e31": {"src": "v14", "tgt": "v16"}}}':

**Comment by jvdwetering** [comment]
That is so weird. I will take a look.

**Comment by jvdwetering** [comment]
How do you produce the second graph from the first? Putting the first in the editor and replacing the H-edges by H-boxes one by one still gives me the correct matrix.

**Comment by RDPEast** [comment]
So I just tried to rewrite the diagram now from scratch and now it is giving me the correct thing. That said this issue has appeared before though I now have no idea what the cause is since this seems to have temporarily resolved itself. I note that in all the times it has appeared it is always accompanied by a zero multiplier to the matrix if that is any clue. Alongside the fact that at the time I saw this it resolved itself (and returned to being problematic) based on whether the hadamard boxes were explicit or wire colours.

I'll keep an eye out for the next time this happens.
